xc/device.ml

```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
+      let fd_of domid = try Some (Hashtbl.find dtof domid) with Not_found -> None
+    end
+
+    module Monitor = struct
+      let create () = Epoll.create1 [Epoll.EPOLL_CLOEXEC]
+      let add m fd = Epoll.ctl m Epoll.EPOLL_CTL_ADD fd [Epoll.EPOLLIN]
+      let remove m fd = Epoll.ctl m Epoll.EPOLL_CTL_DEL fd [Epoll.EPOLLIN]
+      let wait m = Epoll.wait m (Hashtbl.length Lookup.dtof + 1) (-1)
+      let with_event ~ok ~err = function | Epoll.EPOLLIN -> ok() | e -> err e
+    end
+    let m = Monitor.create ()
+
+    let monitor_path domid = Printf.sprintf "/var/run/xen/qmp-event-%d" domid
+    let debug_exn msg e = debug "%s: %s" msg (Printexc.to_string e)
+
+    let remove domid =
+      try
+        match Lookup.fd_of domid with
+        | None -> ()
+        | Some c -> Lookup.remove c domid;
+          Monitor.remove m (Qmp_protocol.to_fd c);
+          Qmp_protocol.close c;
+          debug "Removed QMP Event fd for domain %d" domid
+      with e -> debug_exn (Printf.sprintf "Get exception trying remove qmp on domain-%d" domid) e
+
+    let add domid =
+      try
+        match Lookup.fd_of domid with
+        | Some _ -> ()
+        | None -> let c = Qmp_protocol.connect (monitor_path domid) in
+          Qmp_protocol.negotiate c;
+          Lookup.add c domid;
+          Monitor.add m (Qmp_protocol.to_fd c);
+          debug "Added QMP Event fd for domain %d" domid
+      with e ->
+        debug_exn (Printf.sprintf "QMP domain-%d: negotiation failed: removing socket" domid) e;
+        remove domid
+
+    let wakeup domid =
+        output_string channel_w (Printf.sprintf "%d\n" domid);
+        flush channel_w
+
+    let qmp_event_handle domid qmp_event =
+      (* This function will be extended to handle qmp events *)
+      debug "Get QMP event, domain-%d: %s" domid qmp_event.event
+
+    let qmp_event_thread () =
+      Monitor.add m pipe_r;
+      while true do
+      try
+        let cs = Hashtbl.fold (fun k v acc -> v :: acc) Lookup.dtof [] in
+        let fds = List.map (fun x -> Qmp_protocol.to_fd x, x) cs in
+        let rs = Monitor.wait m in
+        rs |> List.iter (
+          fun (events, fd) ->
+            if fd = pipe_r then
+              events |> List.iter (Monitor.with_event
+                ~ok:(fun () ->
+                       let domid = input_line channel_r in
+                       try add (int_of_string domid)
+                       with e -> debug_exn (Printf.sprintf "Cannot add domain-%s to qmp_event_thread" domid) e)
+                ~err:(fun _ ->
+                        debug "Received unexpected epoll event on pipe_r in qmp_event_thread")
+              )
+            else
+              let c = List.assoc fd fds in
+              match Lookup.domid_of c with
+              | None -> ()
+              | Some domid -> events |> List.iter (Monitor.with_event
+                  ~ok:(fun () ->
+                         if Hashtbl.mem Lookup.dtof domid then try
```

Comment:
Lookup.domid_of fd is already checking for that
~if Hashtbl.mem Lookup.dtof domid then try~